File: Study Plan/Top Interview 150/0290 Word Pattern/WordPattern.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    290. Word Pattern
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/word-pattern/solutions/4678691/c-dictionaries-beats-87-in-runtime-time-and-space-complexity-o-n/
    - Runtime 49 ms (Beats 89.66% of users with C#)
    - Memory 39.48 MB (Beats 8.22% of users with C#)
*/

public class Solution
{
    public bool WordPattern(string pattern, string s)
    {
        var words = s.Split(" ");
        if (pattern.Length != words.Length) return false;

        var letterToWord = new Dictionary<char, string>(pattern.Length);
        var wordToLetter = new Dictionary<string, char>(words.Length);

        for (var index = 0; index < pattern.Length; index++)
        {
            var letter = pattern[index];
            var word = words[index];

            letterToWord.TryAdd(letter, word);
            wordToLetter.TryAdd(word, letter);

            if(letterToWord[letter] != word || wordToLetter[word] != letter)
            {
                return false;
            }
        }

        return true;
    }
}

File: Study Plan/Top Interview 150/0290 Word Pattern/WordPattern.java
/*  
    Problem
    --------------------------------------------------------------------------------
    290. Word Pattern
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 1 ms (Beats 86.93% of users with Java)
    - Memory 41.22 MB (Beats 16.68% of users with Java)
*/

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) return false;
        HashMap<Character,String> letterToWord = new HashMap<>(pattern.length());
        HashMap<String,Character> wordToLetter = new HashMap<>(words.length);
        for (var index = 0; index < pattern.length(); index++) {
            char letter = pattern.charAt(index);
            String word = words[index];
            if (!letterToWord.containsKey(letter)) {
                letterToWord.put(letter, word);
            }
            if (!wordToLetter.containsKey(word)) {
                wordToLetter.put(word, letter);
            }
            if (!letterToWord.get(letter).equals(word)
                || !wordToLetter.get(word).equals(letter)) {
                return false;
            }
        }
        return true;
    }
}